### PR TITLE
feat: implement flow for store code allowed user

### PR DESCRIPTION
**Pull request by bkioshn**
## Describe your changes
Implement flow for store code allowed user

### CODE REVIEW

File: src/lib/pages/codes/components/MyStoredCodesSection.tsx [hunk 1]
```diff
@@ -19,21 +22,34 @@ export const MyStoredCodesSection = ({
...
+        </Heading>
+        <Flex gap={2}>
+          {isMainnet ? (
+            <>
+              <UploadButton />
+              <ProposalButton />
+            </>
+          ) : (
+            <DeployButton />
+          )}
```
Comment (poomthiti) [inline src/lib/pages/codes/components/MyStoredCodesSection.tsx:42]: is this logic only valid for osmosis?

File: src/lib/pages/codes/components/UploadButton.tsx [hunk 2]
```diff
@@ -1,16 +1,41 @@
...
+import { useGovParams } from "lib/services/proposalService";
+import type { Addr } from "lib/types";
 
 export const UploadButton = () => {
+  const { address = "" } = useWallet();
   const navigate = useInternalNavigate();
+  const { data: govParams } = useGovParams();
+  const isAllowed = Boolean(
+    govParams?.uploadAccess?.addresses?.includes(address as Addr)
+  );
```
Comment (poomthiti) [inline src/lib/pages/codes/components/UploadButton.tsx:15]: receive as props, general small component shouldn't be the one to fetch data

File: src/lib/pages/upload/index.tsx [hunk 3]
```diff
@@ -1,19 +1,32 @@
...
+  const navigate = useInternalNavigate();
+  const { data: govParams } = useGovParams();
+  const isAllowed = Boolean(
+    govParams?.uploadAccess?.addresses?.includes(address as Addr)
+  );
 
   useEffect(() => {
-    if (router.isReady) AmpTrack(AmpEvent.TO_UPLOAD);
-  }, [router.isReady]);
+    if (!isAllowed && isMainnet) navigate({ pathname: "/deploy" });
```
Comment (poomthiti) [inline src/lib/pages/upload/index.tsx:27]: add replace option

File: src/lib/pages/upload/index.tsx [hunk 4]
```diff
@@ -1,19 +1,32 @@
...
+import { useGovParams } from "lib/services/proposalService";
+import type { Addr } from "lib/types";
 
 const Upload = () => {
   const router = useRouter();
+  const { isMainnet } = useCurrentNetwork();
+
+  const { address = "" } = useWallet();
+  const navigate = useInternalNavigate();
+  const { data: govParams } = useGovParams();
```
Comment (poomthiti) [inline src/lib/pages/upload/index.tsx:21]: useUploadAccessParams

File: src/lib/pages/codes/components/UploadButton.tsx [hunk 5]
```diff
@@ -1,16 +1,41 @@
...
 
 import { useInternalNavigate } from "lib/app-provider";
 import { CustomIcon } from "lib/components/icon";
+import { useGovParams } from "lib/services/proposalService";
+import type { Addr } from "lib/types";
 
 export const UploadButton = () => {
+  const { address = "" } = useWallet();
   const navigate = useInternalNavigate();
+  const { data: govParams } = useGovParams();
```
Comment (poomthiti) [inline src/lib/pages/codes/components/UploadButton.tsx:12]: useUploadAccessParams

**Review by poomthiti — CHANGES_REQUESTED** [review]

File: src/lib/pages/deploy/index.tsx [hunk 6]
```diff
@@ -85,7 +86,8 @@ const Deploy = () => {
         subtitle="You need to connect wallet to proceed this action"
         mb={4}
       />
-      {address && (
+      {/* Revist isMainnet logic */}
+      {address && isMainnet && (
```
Comment (poomthiti) [inline src/lib/pages/deploy/index.tsx:90]: what if some testnet is permissioned

**Review by poomthiti — CHANGES_REQUESTED** [review]

File: src/lib/pages/upload/index.tsx [hunk 7]
```diff
@@ -1,19 +1,38 @@
...
+  const navigate = useInternalNavigate();
+  const { data, isLoading } = useUploadAccessParams();
+
+  const isPermissionedNetwork =
+    data?.permission !== AccessConfigPermission.EVERYBODY;
+
+  const enableUpload = useMemo(() => {
+    if (!isPermissionedNetwork) return true;
+    return Boolean(data?.addresses?.includes(address as HumanAddr));
+  }, [data, address, isPermissionedNetwork]);
```
[the remark this answers is not in the document]
Comment (poomthiti) [reply src/lib/pages/upload/index.tsx:28]: replace this in deploy index page too thanks 🙏

**Review by evilpeach — APPROVED** [review]

**Pull request merged**